sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -631,6 +631,25 @@ object DataSourceStrategy
     }
   }
 
+  /**
+   * Translates a runtime filter into a data source filter.
+   *
+   * Runtime filters usually contain a subquery that must be evaluated before the translation.
+   * If the underlying subquery hasn't completed yet, this method will throw an exception.
+   */
+  protected[sql] def translateRuntimeFilter(expr: Expression): Option[Filter] = expr match {
+    case in @ InSubqueryExec(e @ PushableColumnAndNestedColumn(name), _, _, _) =>
+      val values = in.values().getOrElse {
+        throw new AnalysisException(s"Can't translate $in to source filter, no subquery result")
```

Comment:
Shall we throw `IllegalStateException`? Seems only bug can lead to this branch.

Reply:
We will only translate the runtime filter when we executing the physical plan, and at that time subqueries must be all evaluated.

Reply:
Makes sense, I'll switch.

Reply:
Fixed.